### DISCUSSION TITLE
Start Easyeffects before having played any audio = silence

### DISCUSSION THREAD

**Discussion by Shallrath** [wwmm/easyeffects#1151] [Q&A]
Hi,

I want to keep easyeffects running in background when I'm logged in. Therefor, I have added easyeffects --gapplication-service to the startup script of my WM (fluxbox). Whenever I log in, easyeffects is running in background, but when I play some audio, I can't hear anything. I need to kill and restart easyeffects and then it works fine.

After some experimenting, I found out that easyeffects must be started **after** some audio has been played to work properly. This is not very nice because it can't be autostarted. Is there a workaround?

Another oddity I noticed is that when starting easyeffects before having played any audio, I can see easyeffects-sink as an output device in the mixer and theVU-meter of it is indicating playback. The VU-meter of my soundcard stays at zero. If I start easyeffects after already having played audio before, easyeffects' VU-meter stays at zero and my soundcard shows a signal. Then I can hear it and I can hear that it is being processed by easyeffects.

I think this might be related that pipewire is started implicitly by a socket when audio is played and that easyeffects is not starting pipewire properly so an audio playback needs to start it first?

**Comment by wwmm** [comment]
>Is there a workaround?

It is the first time I hear about that. In the past a kde user was having a weird problem where kde was initializing our autostart file located at `~/.config/autostart/` way before Pipewire was fully initialized. As a result EasyEffects was loaded in a broken state. Maybe something similar is happening in your fluxbox installation.

In the end we never found out why that was happening to a specific kde user. Others did not have the problem. The solution he found at the time was delaying the autostart. I do not remember how he did that. Is fluxbox fully initializing PipeWire in the login phase or is this happening only on demand when an audio player plays something?

**Reply by Shallrath** [reply]
I added the playback of a short sound to my ~/.fluxbox/startup, followed by a foreground sleep 10 and starting easyeffects aferwards and the problem was still the same. I removed all audio related stuff from startup, started easyeffects manually from a terminal, started some music and it was silent. I killed and restarted easyeffects and then the music was audible.
Fluxbox does not seem to do anything with sound servers, however pactl info seemed to indicate that pipewire is running even before doing any audio-related things.

**Reply by Shallrath** [reply]
Seems that playing an audio or not is not what matters. If I start easyeffects in my startup like this, it works:
```
easyeffects --gapplication-service &
sleep 3
killall easyeffects
easyeffects --gapplication-service &
exec fluxbox
```
So what matters is just to start, kill and start again.

**Reply by Shallrath** [reply]
I compared the log output of easyeffects from the 1st launch when it causes silence and the 2nd launch when it works and there is no difference except for the timestamps.

**Reply by Shallrath** [reply]
What I also tried is adding a `sleep 10` at the beginning to my ~/.fluxbox/startup in case I need to wait a bit for pipewire before doing anything else but this didn't change anything.

**Reply by wwmm** [reply]
In any case show me the logs. Maybe I see something else that will point us to the right direction. Kill any possible instance `easyeffects -q` and restart it in debug mode `G_MESSAGES_DEBUG=easyeffects easyeffects`. It won't be the same as starting in the login but maybe it will help

**Reply by wwmm** [reply]
If you run `easyeffects --gapplication-service` in a terminal does it work? Or does it fail only when executed by the DM in the login?

**Reply by Shallrath** [reply]
Starting easyeffects from terminal or through ~/.fluxbox/startup both gives the same result.

I removed everything audio-related from ~/.fluxbox/startup and did a reboot.

I started easyeffects with debug logging and played some music. Nothing was audible. Pavucontrol showed signal in the easyeffects-sink and not in my soundcard.
[1.log](https://github.com/wwmm/easyeffects/files/7173872/1.log)
![Screenshot 2021-09-16 01h49m39s](https://user-images.githubusercontent.com/69368217/133527952-5f10bd57-f4b2-4649-9acd-71bb8ef63ebd.png)

Then I killed easyeffects, started it again and played music again. Music was audible and the Equalizer of easyeffects was audibly working. Now, the signal appears in my soundcard and not in easyeffects-sink anymore.
[2.log](https://github.com/wwmm/easyeffects/files/7173878/2.log)
![Screenshot 2021-09-16 01h50m28s](https://user-images.githubusercontent.com/69368217/133528036-8412e9db-f142-405e-9777-45323d620381.png)

**Reply by wwmm** [reply]
Looking at your logs I see some problems
```
pipe_manager: new metadata property: 0, default.configured.audio.sink, Spa:String:JSON, { "name": "easyeffects_sink" }
pipe_manager: Audio/Sink 44 alsa_output.pci-0000_01_00.1.hdmi-stereo-extra2 was added
pipe_manager: Audio/Sink 46 alsa_output.pci-0000_00_1f.3.iec958-stereo was added
...
soe:  link from node 67 to output device 4294967295 failed
```
I have seen PipeWire doing this in the past in some users installations. I do not really know why it happens. Before notifying us about your sound card devices PipeWire is telling us that our own virtual device is the default device. In the end what happens is that we do not know where we should send audio. That is why the line starting with `soe` fails. The id `4294967295` is telling that we do not know where to send  audio to.

Something you can try to going to this configuration window
![Screenshot from 2021-09-16 11-40-37](https://user-images.githubusercontent.com/16225172/133632561-ab37dbae-f370-49de-91f2-612e873532ee.png)
, uncheck `Use Default  Output` and select your output device manually. The harder solution would be to try to force PipeWire to somehow load real devices before virtual devices. This way your hardware would probably be the default and we would see this.

**Reply by Shallrath** [reply]
> PipeWire is telling us that our own virtual device is the default device

That was the critical hint. After installing Easyeffects, I wanted to make sure that all applications use this sink instead of the sound card directly, bypassing Easyeffects. So I set the easyeffects-sink to be the fallback device in pavucontrol.

I changed it back for my sound card to be the fallback device and now Easyeffects is working correctly when only started once. Verifyed after a reboot.

@wwmm Thank you very much for your immediate and patient support and of course for Easyeffects :1st_place_medal:

**Reply by wwmm** [reply]
Oh! I see. I am glad it is working now. Lots of people do that. But in reality EasyEffects(and PulseEffects) is designed assuming that your hardware is the default device. The only cases where you really need to set our virtual devices as default is when you use audio applications that ask PipeWire/Pulseaudio to not allow their streams to be managed. But as you saw it isn't a workaround for everybody. Sometimes it happens what happened to you.

**Comment by wwmm** [comment]
>I think this might be related that pipewire is started implicitly by a socket when audio is played and that easyeffects is not starting pipewire properly so an audio playback needs to start it first?

At least on GNOME PipeWire is fully working by the time my login finishes. If there is some kind of socket activation Gnome is forcing it to happen very early.
